core/src/main/scala/kafka/controller/KafkaController.scala

```diff
@@ -664,41 +646,40 @@ class KafkaController(val config: KafkaConfig,
-    * @return An option of the original replicas prior to the ongoing reassignment. None if there is no ongoing reassignment
-    */
-  private def maybeRevertOngoingReassignment(topicPartition: TopicPartition, reassignedPartitionContext: ReassignedPartitionsContext): Option[Seq[Int]] = {
-    reassignedPartitionContext.ongoingReassignmentOpt match {
-      case Some(ongoingAssignment) =>
-        val originalAssignment = ongoingAssignment.previousAssignment
-        assert(ongoingAssignment.isBeingReassigned)
-        assert(!originalAssignment.isBeingReassigned)
+   * Update the current assignment state in Zookeeper and in memory. If a reassignment is already in
+   * progress, then the new reassignment will supplant it and some replicas will be shutdown.
+   *
+   * Note that due to the way we compute the original replica set, we have no guarantee that a revert
+   * would put it in the same order.
+   *
+   * @param topicPartition The reassigning partition
+   * @param reassignment The new reassignment context
+   *
+   * @return The updated assignment state
+   */
+  private def updateCurrentReassignment(topicPartition: TopicPartition,
+                                        reassignment: ReassignmentContext): ReplicaAssignment = {
+    val currentAssignment = controllerContext.partitionFullReplicaAssignment(topicPartition)
+    val updatedAssignment = currentAssignment.reassignTo(reassignment.newReplicas)
 
-        val unnecessaryReplicas = ongoingAssignment.replicas.filterNot(originalAssignment.replicas.contains(_))
-        val (overlappingReplicas, replicasToRemove) = unnecessaryReplicas.partition(reassignedPartitionContext.newReplicas.contains(_))
-        // RS = ORS + OVRS, AR = OVRS, RR = []
-        val intermediateReplicaAssignment = PartitionReplicaAssignment(originalAssignment.replicas ++ overlappingReplicas, overlappingReplicas, Seq())
+    if (currentAssignment != updatedAssignment) {
+      info(s"Updated assignment of partition $topicPartition from $currentAssignment to $updatedAssignment")
 
-        if (isDebugEnabled)
-          debug(s"Reverting previous reassignment $originalAssignment (we were in the " +
-            s"process of an ongoing reassignment with target replicas ${ongoingAssignment.targetReplicas.mkString(",")} (" +
-            s"Overlapping replicas: ${overlappingReplicas.mkString(",")}, Replicas to remove: ${replicasToRemove.mkString(",")})")
+      // A1 and A2: Update zk and cached state
+      updateReplicaAssignmentForPartition(topicPartition, updatedAssignment)
+      controllerContext.updatePartitionFullReplicaAssignment(topicPartition, updatedAssignment)
 
-        // Set RS = ORS + OVRS, AR = OVRS, RR = [] in memory.
-        controllerContext.updatePartitionFullReplicaAssignment(topicPartition, intermediateReplicaAssignment)
+      // If there is a reassignment already in progress, then some of the currently adding replicas
```

Comment:
It's a good question. I think we have this problem for replica removal in general even prior to KIP-455. The StopReplica requests are all sent asynchronously and we do not await their completion as we do with topic deletion. I think this will be fixed by KIP-500.

Reply:
Oh. Sounds good then